docs/developer-docs/security/security-best-practices/data-integrity-and-authenticity.mdx

```diff
@@ -8,19 +8,334 @@ import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";
+    }
+}
+
+#[derive(CandidType)]
+struct CertifiedUser {
+    user: User,
+    certificate: Vec<u8>,
+    witness: Vec<u8>,
+}
+
+thread_local! {
+    static INDEX : Cell<u64> = Cell::new(0);
+    static TREE: RefCell<RbTree<&'static str, RbTree<[u8; 8], User>>> = RefCell::new(RbTree::new());
+}
+
+#[ic_cdk::update]
+fn set_user(user: User) -> u64 {
+    let index = INDEX.with(|index| {
+        let count = index.get() + 1;
+        index.set(count);
+        count
+    });
+
+    TREE.with_borrow_mut(|tree| {
+        match tree.get(b"user") {
+            Some(_) => {
+                tree.modify(b"user", |inner| {
+                    inner.insert(index.to_be_bytes(), user);
+                });
+            }
+            None => {
+                let mut inner = RbTree::new();
+                inner.insert(index.to_be_bytes(), user);
+                tree.insert("user", inner);
+            }
+        }
+        ic_cdk::api::set_certified_data(&tree.root_hash());
+    });
+    index
+}
+
+#[ic_cdk::query]
+fn get_user(index: u64) -> CertifiedUser {
+    let certificate = ic_cdk::api::data_certificate().expect("No data certificate available");
+
+    TREE.with_borrow(|tree| {
+        let user = match tree.get(b"user") {
+            Some(inner) => {
+                let user = inner.get(&index.to_be_bytes()[..]).expect("User not found");
+                user.to_owned()
+            }
+            None => {
+                panic!("Tree isn't initialized");
+            }
+        };
+
+        let mut witness = vec![];
+        let mut witness_serializer = serde_cbor::Serializer::new(&mut witness);
+        let _ = witness_serializer.self_describe();
+        tree.nested_witness(b"user", |inner| inner.witness(&index.to_be_bytes()[..]))
+            .serialize(&mut witness_serializer)
+            .unwrap();
+
+        CertifiedUser {
+            user,
+            certificate,
+            witness,
+        }
+    })
+}
+```
+
+### Verifying the Certified Variables
+
+Once you have the response `CertifiedUser`, for the secureness guarantee, one must verify the certification in the response. This is broken down into several steps with an example below implementing these steps in Rust.
+
+:::note
+The example has some extra steps to setup the canister with some `User` data before verification. You can ignore the section marked between `// ==== START of canister data setup` and `// ==== END of canister data setup`
+:::
+
+1. Recompute the `root_hash` of `certificate.tree` (pruned state tree with the canister's `certified_data`) and verify the `certificate.signature` with `root_hash` as message and `certificate.delegation` and IC root_key as the public key. This confirms that the signature is valid for the current state tree. 
+2. Validate that the response is not stale by verifying the time at `/time` in `certificate.tree` is less than a certain delta of currrent time. The recommended delta is 5 minutes.
```

Comment:
```suggestion
2. Validate that the response is not stale by verifying the time at `/time` in `certificate.tree` is less than a certain delta of current time. The recommended delta is 5 minutes but should be adapted to the use case.
```
I'm wondering if we have a concrete example where the /time would have to be more recent than 5 minutes. Maybe some sort of DeFi app where price information determines the default order price which could trick users into using a 5 min old value that would make them overpay on a market order.

Reply:
We could add one when we add a best practice about system time. Right now, even agents enforce 5m and not sure if it's adaptable.